python/tank/platform/interactive_authentication.py

```diff
@@ -0,0 +1,195 @@
+    print "Please enter your password to renew your session for %s" % hostname
+    password = getpass()
+    return password
+
+
+def _do_console_based_login(hostname, default_login, http_proxy):
+    """
+    Logs to the Shotgun site using username/password combo. Retries if the
+    login attempt failed. Everything is done on the command line.
+    :param default_hostname: Default value for the hostname.
+    :param default_login: Default value for the login.
+    :param http_proxy: Proxy to use when connecting.
+    :returns: A valid Shotgun instance.
+    """
+    while True:
+        # Get the credentials from the user
+        login, password = _get_user_credentials_from_keyboard(
+            hostname,
+            default_login
+        )
+        session_token = _get_session_token(hostname, login, password, http_proxy)
+        if session_token:
+            return session_token, login
+
+
+def _do_ui_based_session_renewal(hostname, login, http_proxy):
+    pass
+
+
+def _do_console_based_session_renewal(hostname, login, http_proxy):
+    """
+    Prompts the user for this password to retrieve a new session token and rewews
+    the session token.
+    :param hostname: Host to renew a token for.
+    :param login: User to renew a token for.
+    :param http_proxy: Proxy to use for the request. Can be None.
+    :returns: The (session token, login user) tuple.
+    """
+    while True:
+        # Get the credentials from the user
+        password = _get_user_pass_from_keyboard(hostname, login)
+        session_token = _get_session_token(hostname, login, password, http_proxy)
+        if session_token:
+            return session_token, login
+
+
+def _do_logging(login_functor):
+    """
+    Common login logic, regardless of how we are actually logging in. It will first try to reuse
+    any existing session and if that fails then it will ask for credentials and upon success
+    the credentials will be cached.
+    :param login_functor: Functor that gets invoked to retrieve the credentials of the user.
+    """
+    if authentication.is_authenticated():
+        return
+
+    config_data = shotgun.get_associated_sg_config_data()
+    # We might not have login information, in that case use an empty dictionary.
+    login_info = authentication.get_cached_login_info(config_data["host"]) or {}
+
+    # Ask for the credentials
+    session_token, login = login_functor(
+        config_data["host"],
+        login_info.get("login", get_login_name()),
+        config_data.get("http_proxy")
+    )
+
+    logger.debug("Login successful!")
+
+    # Cache the credentials so subsequent session based logins can reuse the session id.
+    authentication.cache_session_data(config_data["host"], login, session_token)
+
+
+def console_renew_session():
+    """
+    Prompts the user to enter his password on the command line to retrieve a new session token.
+    """
+    _do_logging(_do_ui_based_session_renewal)
```

Comment:
So this construct feels a little confusing for me. Wondering if it could be something like this?

```python
def console_renew_session():
    """
    Prompts the user to enter his password on the command line to retrieve a new session token.
    """
    if authentication.is_authenticated():
        # all good!
        return
    # get authentication details, host, user name etc.
    c = _get_credentials()
    # do the actual renewal
    (token, login) = _do_ui_based_session_renewal(c)
    # Cache the credentials so subsequent session based logins can reuse the session id.
    authentication.cache_session_data(c, login, session_token)
```

I realise that this means a little bit of repetition in the methods, but I am wondering if that may make things a little bit clearer?